easybuild/easyblocks/p/pytorch.py

```diff
@@ -266,73 +267,95 @@ def test_step(self):
-
-        # Loop over the second pattern to count failures/errors
-        for summary in summary_matches_pattern2:
-            failures = get_count_for_pattern(r"^.*[^0-9](?P<failures>[0-9]+) failed.*$", summary[0])
-            failure_cnt += failures
-            errs = get_count_for_pattern(r"^.*[^0-9](?P<errors>[0-9]+) error.*$", summary[0])
-            error_cnt += errs
-
-        # Calculate total number of unsuccesful tests
+        failed_test_suites = []
+
+        # Grep for patterns like:
+        # Ran 219 tests in 67.325s
+        #
+        # FAILED (errors=10, skipped=190, expected failures=6)
+        # test_fx failed!
+        regex = (r"^Ran (?P<test_cnt>[0-9]+) tests.*$\n\n"
+                 r"FAILED \((?P<failure_summary>.*)\)$\n"
+                 r"(?:^(?:(?!failed!).)*$\n)*"
+                 r"(?P<failed_test_suite_name>.*) failed!(?: Received signal: \w+)?\s*$")
+
+        for m in re.finditer(regex, tests_out, re.M):
+            # E.g. 'failures=3, errors=10, skipped=190, expected failures=6'
+            failure_summary = m['failure_summary']
+            total, test_suite = m.group('test_cnt', 'failed_test_suite_name')
+            failure_report += "{test_suite} ({total} total tests, {failure_summary})\n".format(
+                    test_suite=test_suite, total=total, failure_summary=failure_summary
+                )
+            failure_cnt += get_count_for_pattern(r"(?<!expected )failures=([0-9]+)", failure_summary)
+            error_cnt += get_count_for_pattern(r"errors=([0-9]+)", failure_summary)
+            failed_test_suites.append(test_suite)
+
+        # Grep for patterns like:
+        # ===================== 2 failed, 128 passed, 2 skipped, 2 warnings in 3.43s =====================
+        regex = r"^=+ (?P<failure_summary>.*) in [0-9]+\.*[0-9]*[a-zA-Z]* =+$\n(?P<failed_test_suite_name>.*) failed!$"
+
+        for m in re.finditer(regex, tests_out, re.M):
+            # E.g. '2 failed, 128 passed, 2 skipped, 2 warnings'
+            failure_summary = m['failure_summary']
+            test_suite = m['failed_test_suite_name']
+            failure_report += "{test_suite} ({failure_summary})\n".format(
+                    test_suite=test_suite, failure_summary=failure_summary
+                )
+            failure_cnt += get_count_for_pattern(r"([0-9]+) failed", failure_summary)
+            error_cnt += get_count_for_pattern(r"([0-9]+) error", failure_summary)
+            failed_test_suites.append(test_suite)
+
+        # Make the names unique
+        failed_test_suites = nub(failed_test_suites)
+        # Gather all failed tests suites in case we missed any (e.g. when it exited due to syntax errors)
+        all_failed_test_suites = nub(
+            re.findall(r"^(?P<test_name>.*) failed!(?: Received signal: \w+)?\s*$", tests_out, re.M)
+        )
+        # If we missed any test suites prepend a list of all failed test suites
+        if failed_test_suites != all_failed_test_suites:
```

Comment:
These two will not be the same in the case where we have a failed test that matches the second pattern before one that matches the first pattern. I think we can just change the comparison to be

```py
if sorted(failed_test_suites) != sorted(all_failed_test_suites):
```

Example:

So for
```
Ran 219 tests in 67.325s

FAILED (errors=10, skipped=190, expected failures=6)
test_p1 failed!

===================== 2 failed, 128 passed, 2 skipped, 2 warnings in 3.43s =====================
test_p2 failed!
```
`failure_report` is
```
test_p1 (219 total tests, errors=10, skipped=190, expected failures=6)
test_p2 (2 failed, 128 passed, 2 skipped, 2 warnings)
```

But for
```
===================== 2 failed, 128 passed, 2 skipped, 2 warnings in 3.43s =====================
test_p2 failed!

Ran 219 tests in 67.325s

FAILED (errors=10, skipped=190, expected failures=6)
test_p1 failed!
```
`failure_report` is
```
Failed tests (suites/files):
* test_p1
* test_p2
test_p1 (219 total tests, errors=10, skipped=190, expected failures=6)
test_p2 (2 failed, 128 passed, 2 skipped, 2 warnings)
```

Reply:
I guess it makes sense to not use `nub` but simply `sorted(set`. Added that as a commit.

Note that `nub` creates a set **and** a list so this might actually be faster.